drivers/clock_control/clock_control_silabs_siwx917.c

```diff
@@ -0,0 +1,197 @@
+	uintptr_t clockid = (uintptr_t)sys;
+
+	switch (clockid) {
+	case SIWX917_CLK_ULP_I2C:
+		RSI_ULPSS_PeripheralDisable(ULPCLK, ULP_I2C_CLK);
+		break;
+	case SIWX917_CLK_ULP_DMA:
+		RSI_ULPSS_PeripheralDisable(ULPCLK, ULP_UDMA_CLK);
+		break;
+	case SIWX917_CLK_UART1:
+		RSI_CLK_PeripheralClkDisable(M4CLK, USART1_CLK);
+		break;
+	case SIWX917_CLK_UART2:
+		RSI_CLK_PeripheralClkDisable(M4CLK, USART2_CLK);
+		break;
+	case SIWX917_CLK_DMA0:
+		RSI_CLK_PeripheralClkDisable(M4CLK, UDMA_CLK);
+		break;
+	case SIWX917_CLK_ULP_UART:
+	case SIWX917_CLK_I2C0:
+	case SIWX917_CLK_I2C1:
+		/* Not supported */
+		return 0;
+	default:
+		return -EINVAL;
+	}
+
+	data->enable &= ~BIT(clockid);
+	return 0;
+}
+
+static int siwx917_clock_get_rate(const struct device *dev, clock_control_subsys_t sys,
+				  uint32_t *rate)
+{
+	uintptr_t clockid = (uintptr_t)sys;
+
+	switch (clockid) {
+	case SIWX917_CLK_ULP_UART:
+		*rate = 32000000;
+		return 0;
+	case SIWX917_CLK_UART1:
+		*rate = 25000000;
+		return 0;
+	case SIWX917_CLK_UART2:
+		*rate = 20000000;
+		return 0;
+	default:
+		/* For now, no other driver need clock rate */
+		return -EINVAL;
+	}
+}
+
+static enum clock_control_status siwx917_clock_get_status(const struct device *dev,
+							  clock_control_subsys_t sys)
+{
+	struct siwx917_clock_data *data = dev->data;
+	uintptr_t clockid = (uintptr_t)sys;
+
+	if (data->enable & BIT(clockid)) {
+		return CLOCK_CONTROL_STATUS_ON;
+	} else {
+		return CLOCK_CONTROL_STATUS_OFF;
+	}
+}
+
+static int siwx917_clock_init(const struct device *dev)
+{
+	SystemCoreClockUpdate();
+
+	/* Use SoC PLL at configured frequency as core clock */
+	sl_si91x_clock_manager_m4_set_core_clk(M4_SOCPLLCLK, CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+
+	/* Use interface PLL at configured frequency as peripheral clock */
+	sl_si91x_clock_manager_set_pll_freq(INFT_PLL, CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,
+					    PLL_REF_CLK_VAL_XTAL);
+
+	/* FIXME: Currently the clock consumer use clocks without power on them.
+	 * This should be fixed in drivers. Meanwhile, get the list of required
+	 * clocks using DT labels.
+	 */
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(ulpuart), okay)
+	siwx917_clock_on(dev, (clock_control_subsys_t)SIWX917_CLK_ULP_UART);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(ulpi2c), okay)
+	siwx917_clock_on(dev, (clock_control_subsys_t)SIWX917_CLK_ULP_I2C);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart1), okay)
+	siwx917_clock_on(dev, (clock_control_subsys_t)SIWX917_CLK_UART1);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart2), okay)
+	siwx917_clock_on(dev, (clock_control_subsys_t)SIWX917_CLK_UART2);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(i2c0), okay)
+	siwx917_clock_on(dev, (clock_control_subsys_t)SIWX917_CLK_I2C0);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(i2c1), okay)
+	siwx917_clock_on(dev, (clock_control_subsys_t)SIWX917_CLK_I2C1);
+#endif
+
+	return 0;
+}
+
+static const struct clock_control_driver_api siwx917_clock_api = {
```

Comment:
I believe this should be using `DEVICE_API()` like here:
https://github.com/zephyrproject-rtos/zephyr/blob/59c94db6f052c9d848b4cec2008152d32d8ee597/drivers/clock_control/clock_control_nrf.c#L695-L700

It's a fairly new convention, so if we're not pointing at a recent enough upstream yet, then that needs to be fixed first. Most likely the same update will need to be made to any other of our downstream 917 drivers.